### ISSUE TITLE
Line graph raises an error when nodes are not ordered.

### ISSUE THREAD

**Issue by dschult** [networkx/networkx#5329]
When nodes are not ordered, the line_graph function raises a TypeError.
```python
G = nx.path_graph([0, 1, 2, "a", "b", "c"])
nx.line_graph(G)
```

This happens because we want the 2-tuple identity of an edge to become a node in the line_graph. So to ensure a unique representation, we rewrite all edge tuples in the canonical order (u, v) where u<v.  There are other ways to achieve this outcome that don't rely on ordered nodes.  Let's use one of them. :}

For example, we can map the nodes to integers:  `node_index = {n: i for i, n in enumerate(G)}`
and then order the nodes by comparing the `node_index` values.

**Comment by jkudla** [comment]
Hi, can I work on this issue? I'm new to the project and followed the first steps :)

Idea would be introducing the `node_index` mapping as you described, @dschult. Run the same construction, perhaps with something like `nodes = [tuple(sorted([node_index[x[0]], node_index[x[1]]])) + x[2:] for x in get_edges(u)]`. Finally, transform the node indices back to the original node labels. This could be done using a reverse mapping `index_node` from indices back to nodes. For now, I implemented this by first constructing `L` and then building a graph `L_result` which is isomorphic to `L` but has the original labels on its nodes and edges. Any thoughts about this procedure?

I did tried some approaches and suggest the above mainly because directly incorporating the original node labels as we construct `L` seems shorter, but it's harder to make this correct as we want to rely on canonical forms for nodes and edges. Even "enriching" the nodes with their additional labels seems to make the code rather complicated.

Note that I haven't finished a PR yet since multigraphs are still an issue.

**Comment by dschult** [comment]
Thanks for looking at this.  I think the best approach is to leave the node labels as they are, but any time they are sorted, use the index dict as the "key" function for the sorting: `sorted(x[:2], key=node_index.__get__)`

**Comment by jkudla** [comment]
Thank you for pointing this out, that's much more elegant!

**Comment by dschult** [comment]
Fixes by #5357
